internal/manager/pvcleanup/controller.go

```diff
@@ -0,0 +1,227 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package pvcleanup
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"local-csi-driver/internal/csi/core/lvm"
+)
+
+const (
+	// PVProtectionFinalizer is the standard Kubernetes finalizer that prevents PV deletion
+	PVProtectionFinalizer = "kubernetes.io/pv-protection"
+
+	// ExternalProvisionerFinalizer is the finalizer set by the external provisioner
+	ExternalProvisionerFinalizer = "external-provisioner.volume.kubernetes.io/finalizer"
+
+	// LocalCSIDriverName is the name of our CSI driver
+	LocalCSIDriverName = "localdisk.csi.acstor.io"
+)
+
+// PVCleanupReconciler watches PV delete events and removes finalizer if node doesn't exist
+type PVCleanupReconciler struct {
+	client.Client
+	Recorder record.EventRecorder
+}
+
+// Reconcile handles PV reconciliation
+func (r *PVCleanupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	// Fetch the PV
+	var pv corev1.PersistentVolume
+	if err := r.Get(ctx, req.NamespacedName, &pv); err != nil {
+		if apierrors.IsNotFound(err) {
+			// PV was deleted, nothing to do
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "failed to get PersistentVolume")
+		return ctrl.Result{}, err
+	}
+
+	// Only process PVs from our CSI driver
+	if pv.Spec.CSI == nil || pv.Spec.CSI.Driver != lvm.DriverName {
+		logger.V(4).Info("PV is not from our CSI driver, skipping", "pv", pv.Name, "driver", pv.Spec.CSI)
+		return ctrl.Result{}, nil
+	}
+
+	// Only process PVs that are Released with Delete reclaim policy
+	if pv.Status.Phase != corev1.VolumeReleased {
+		// PV is not released, nothing to do
+		return ctrl.Result{}, nil
+	}
+
+	if pv.Spec.PersistentVolumeReclaimPolicy != corev1.PersistentVolumeReclaimDelete {
+		// PV doesn't have Delete reclaim policy, nothing to do
+		logger.V(4).Info("PV doesn't have Delete reclaim policy, skipping", "pv", pv.Name, "reclaimPolicy", pv.Spec.PersistentVolumeReclaimPolicy)
+		return ctrl.Result{}, nil
+	}
+
+	hasPVProtection := hasFinalizer(&pv, PVProtectionFinalizer)
```

Comment:
controller-runtime has `ContainersFinalizer` that can be used

https://github.com/kubernetes-sigs/controller-runtime/blob/dd11b6f36e94cf95f8d8875617382780fc51af40/pkg/controller/controllerutil/controllerutil.go#L531